### PR TITLE
feat: add bootstrap/listener/cluster/route configdump support

### PR DESCRIPTION
**Pull request by Xunzhuo**
xref: https://github.com/envoyproxy/gateway/issues/915

### CODE REVIEW

**Comment by Xunzhuo** [comment]
cc @zirain

File: internal/cmd/egctl/config.go [hunk 1]
```diff
@@ -11,10 +11,14 @@ import (
 	"net/http"
 
 	"github.com/spf13/cobra"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/reflect/protoreflect"
 	"k8s.io/apimachinery/pkg/types"
 	cmdutil "k8s.io/kubectl/pkg/cmd/util"
 	"sigs.k8s.io/yaml"
```
Comment (zirain) [inline internal/cmd/egctl/config.go:19]: nit: useless empty line
why lint pass?

File: internal/cmd/egctl/config_test.go [hunk 2]
```diff
@@ -70,38 +70,134 @@ func (fw *fakePortForwarder) Address() string {
...
 	for _, tc := range cases {
 		t.Run(tc.output, func(t *testing.T) {
-			got, err := extractConfigDump(fw, tc.output)
+			configDump, err := extractConfigDump(fw)
+			assert.NoError(t, err)
+			resource, err := findXDSResourceFromConfigDump(tc.resourceType, configDump)
 			assert.NoError(t, err)
-			assert.Equal(t, string(readConfig(tc.expected)), string(got))
+			got, err := marshalEnvoyProxyConfig(resource, tc.output)
+			t.Log("GOT:-----------------\n\n", string(got), "\n\n\n")
```
Comment (zirain) [inline internal/cmd/egctl/config_test.go:174]: remote this

File: internal/cmd/egctl/config.go [hunk 3]
```diff
@@ -111,7 +119,288 @@ func runAllConfig(c *cobra.Command, args []string) error {
...
+
+	_, err = fmt.Fprintln(c.OutOrStdout(), string(out))
+	return err
+}
+
+func bootstrapConfigCmd() *cobra.Command {
+
+	configCmd := &cobra.Command{
+		Use:   "bootstrap <pod-name>",
+		Short: "Retrieves bootstrap Envoy xDS resources from the specified pod",
```
Comment (zirain) [inline internal/cmd/egctl/config.go:140]: alias with `b` for short?

File: internal/cmd/egctl/config.go [hunk 4]
```diff
@@ -111,7 +119,288 @@ func runAllConfig(c *cobra.Command, args []string) error {
...
+	if err := fw.Start(); err != nil {
+		return err
+	}
+	defer fw.Stop()
+
+	configDump, err := extractConfigDump(fw)
+	if err != nil {
+		return err
+	}
+
```
Comment (zirain) [inline internal/cmd/egctl/config.go:190]: can we reuse this part?

File: internal/cmd/egctl/config.go [hunk 5]
```diff
@@ -111,7 +119,288 @@ func runAllConfig(c *cobra.Command, args []string) error {
...
+	_, err = fmt.Fprintln(c.OutOrStdout(), string(out))
+	return err
+}
+
+func listenerConfigCmd() *cobra.Command {
+
+	configCmd := &cobra.Command{
+		Use:   "listener <pod-name>",
+		Short: "Retrieves listener Envoy xDS resources from the specified pod",
```
Comment (zirain) [inline internal/cmd/egctl/config.go:278]: add alias `l`

File: internal/cmd/egctl/config.go [hunk 6]
```diff
@@ -111,7 +119,288 @@ func runAllConfig(c *cobra.Command, args []string) error {
...
+	}
+
+	_, err = fmt.Fprintln(c.OutOrStdout(), string(out))
+	return err
+}
+
+func routeConfigCmd() *cobra.Command {
+
+	configCmd := &cobra.Command{
+		Use:   "route <pod-name>",
```
Comment (zirain) [inline internal/cmd/egctl/config.go:346]: add alias `r`

File: internal/cmd/egctl/config.go [hunk 7]
```diff
@@ -111,7 +119,288 @@ func runAllConfig(c *cobra.Command, args []string) error {
...
+	}
+
+	_, err = fmt.Fprintln(c.OutOrStdout(), string(out))
+	return err
+}
+
+func clusterConfigCmd() *cobra.Command {
+
+	configCmd := &cobra.Command{
+		Use:   "cluster <pod-name>",
```
Comment (zirain) [inline internal/cmd/egctl/config.go:208]: add alias `c`

File: internal/cmd/egctl/config.go (hunk 1 above)
Comment (Xunzhuo) [reply internal/cmd/egctl/config.go:19]: which line is useless? 👀

File: internal/cmd/egctl/config_test.go (hunk 2 above)
Comment (Xunzhuo) [reply internal/cmd/egctl/config_test.go]: Done. Thanks.

File: internal/cmd/egctl/config.go (hunk 3 above)
Comment (Xunzhuo) [reply internal/cmd/egctl/config.go]: Done. Thanks.

File: internal/cmd/egctl/config.go (hunk 5 above)
Comment (Xunzhuo) [reply internal/cmd/egctl/config.go]: Done. Thanks.

File: internal/cmd/egctl/config.go (hunk 6 above)
Comment (Xunzhuo) [reply internal/cmd/egctl/config.go]: Done. Thanks.

File: internal/cmd/egctl/config.go (hunk 7 above)
Comment (Xunzhuo) [reply internal/cmd/egctl/config.go]: Done. Thanks.

File: internal/cmd/egctl/config.go (hunk 4 above)
Comment (Xunzhuo) [reply internal/cmd/egctl/config.go]: Done. Thanks

**Comment by Xunzhuo** [comment]
Done @zirain, PTAL.

File: internal/cmd/egctl/testdata/config/out/out.cluster.json [hunk 8]
```diff
@@ -0,0 +1 @@
.../envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext", "commonTlsContext":{"tlsParams":{"tlsMaximumProtocolVersion":"TLSv1_3"}, "tlsCertificateSdsSecretConfigs":[{"name":"xds_certificate", "sdsConfig":{"pathConfigSource":{"path":"/sds/xds-certificate.json"}, "resourceApiVersion":"V3"}}], "validationContextSdsSecretConfig":{"name":"xds_trusted_ca", "sdsConfig":{"pathConfigSource":{"path":"/sds/xds-trusted-ca.json"}, "resourceApiVersion":"V3"}}}}}}, "lastUpdated":"2023-02-23T09:05:23.436Z"}], "dynamicActiveClusters":[{"versionInfo":"2a0a1698a9d3e05b802047b0cd36b52a070afa49042e1ba267168c5265c7cabf", "cluster":{"@type":"type.googleapis.com/envoy.config.cluster.v3.Cluster", "name":"default-backend-rule-0-match-0-www.example.com", "type":"STATIC", "connectTimeout":"5s", "loadAssignment":{"clusterName":"default-backend-rule-0-match-0-www.example.com", "endpoints":[{"locality":{}, "lbEndpoints":[{"endpoint":{"address":{"socketAddress":{"address":"10.96.159.190", "portValue":3000}}}, "loadBalancingWeight":1}], "loadBalancingWeight":1}]}, "dnsLookupFamily":"V4_ONLY", "outlierDetection":{}, "commonLbConfig":{"localityWeightedLbConfig":{}}}, "lastUpdated":"2023-02-23T09:05:38.443Z"}]}
```
Comment (zirain) [inline internal/cmd/egctl/testdata/config/out/out.cluster.json:1]: IIRC, default wirter is `stdout`, so maybe marshal json with indent?

Comment (Xunzhuo) [reply internal/cmd/egctl/testdata/config/out/out.cluster.json]: Done @zirain

**Review by zirain — APPROVED** [review]
LGTM, thanks!

**Pull request merged**